Implement the Sales Class

rdeleon-bran opened the pull request:
## Describe your changes (Include screenshots if neccessary)

## Issue ticket number and link
Closes #10 
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

rdeleon-bran committed:
Add files via upload

BaoNguyen09 reviewed (changes requested):
Can you put these two files inside [this folder](https://github.com/BaoNguyen09/csc_335_restaurant/tree/main/csc_335_restaurant/src): ?

BaoNguyen09 commented on Sales.java:
```diff
@@ -0,0 +1,52 @@
+	 * and adding the quantity
+	 *  */
+	public void addCompletedOrder(Bill bill) {
+		for (FoodData food: bill.getOrder()) {
```
Currently, there's no `getOrder` method in Bill, so i think you will have to add it if you want to use it

rdeleon-bran replied:
uploaded the Bill class with the getOrder method.

BaoNguyen09 commented on Sales.java:
```diff
@@ -0,0 +1,52 @@
+	 *  */
+	public void addCompletedOrder(Bill bill) {
+		for (FoodData food: bill.getOrder()) {
+			if(!(sales.keySet().contains(food.getName()))) {
+				sales.put(food.getName(), food.getQuantity());
+				continue;
+			}
+			sales.put(food.getName(), sales.get(food.getName()) + food.getQuantity());
+		}
+	}
```
To ensure no unnecessary duplication, we should allow case-insensitive food name
```suggestion
			String foodName = food.getName().toLowerCase();
			if(!(sales.keySet().contains(foodName))) {
				sales.put(foodName, food.getQuantity());
				continue;
			}
			sales.put(foodName, sales.get(foodName) + food.getQuantity());
		}
	}
```

BaoNguyen09 commented on Sales.java:
This class needs at least two more methods to satisfy the spec:
- Sort sales based on most frequently ordered items.
- Sort sales based on the total amount of money made from a specific item – i.e. the cost per item * the number of items sold

And in order to calculate the amount of money, it will need to use Menu to search for price, which means the HashMap should use FoodData type to be able to trace the amount of the food

rdeleon-bran replied:
added them and made it so that the key of the hashmaps is FoodData type.

rdeleon-bran committed:
Made changes to the sales class

changed the hashcode and equals method in the foodData class. along with edits to the bill class and sales class.

rdeleon-bran committed:
Delete Sales.java

not in right folder

rdeleon-bran committed:
Delete SalesTests.java

not in right folder

rdeleon-bran committed:
Delete csc_335_restaurant/src/Bill.java

rdeleon-bran committed:
Delete csc_335_restaurant/src/BillTests.java

rdeleon-bran committed:
Delete csc_335_restaurant/src/FoodData.java

rdeleon-bran committed:
Delete csc_335_restaurant/src/SalesTests.java

rdeleon-bran committed:
Delete csc_335_restaurant/src/Sales.java

rdeleon-bran committed:
Add updated Sales, FoodData, and Bill in model

updated Sales to match the specs, changes hashmaps to FoodData as the key. Updated the Bill class to add getOrder. Updated the FoodData class so the equal method works.

rdeleon-bran committed:
Added the updated tests

Did not update the foodData tests so if equals method was tested in the original then it might not work.

rdeleon-bran committed:
Merge branch 'main' into implement-Sales-Class

rdeleon-bran committed:
Update FoodData tests

rdeleon-bran reviewed (commented):
Reviewed all of the files and made sure they match my local code.

BaoNguyen09 commented on csc_335_restaurant/src/model/Bill.java:
```diff
@@ -52,6 +61,11 @@ public double calculateBill() {
 	}
 	
+	public ArrayList<FoodData> getOrder(){
+		ArrayList<FoodData> copyOrder = new ArrayList<>(this.order);
```
Currently, this code just returns a shallow copy of the order, and because FoodData is mutable that will cause escaping reference issue.
```suggestion
		ArrayList<FoodData> copyOrder = new ArrayList<>();
                for (FoodData fd : this.order) {
                    copyOrder.add(new FoodData(fd));  // new instance each time
                }
```

BaoNguyen09 commented on csc_335_restaurant/src/model/FoodData.java:
Can you add a copy constructor for this class so that we can use it to create deep copy?

BaoNguyen09 commented on csc_335_restaurant/src/model/FoodData.java:
```diff
@@ -45,26 +46,17 @@ public double getPrice() {
-			return false;
-		FoodData other = (FoodData) obj;
-		return Objects.equals(modifications, other.modifications) && quantity == other.quantity;
+		return super.equals(obj);
```
Why do you have to change it to this?

rdeleon-bran replied:
changed it so that contains in addCompletedOrder() works properly. The goal of the hashMap is to keep track of how many food items are added. Before i changed it if the quantity or modifications was different then the key would be added separately from the original food despite having the same name, type, and price.

BaoNguyen09 replied:
Than instead of using FoodData as key, I think it makes more sense to use Food as the key since it's immutable

rdeleon-bran replied:
Yes it would make more sense but the FoodData class does not have a getFood method. There also does not seem to be a problem with changing the equals method to just match the food properties. If the equals method in FoodData matched depending on quantity and modifications too then it would be very unlikely anything matches.

BaoNguyen09 replied:
Then you make the FoodData in Sales act as Food, so i think it's better to use Food as the key instead of FoodData

BaoNguyen09 replied:
This should also be reverted

rdeleon-bran replied:
reverted it

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,92 @@
+
 
+public class Sales {
+	/* I made sales HashMap String, Integer instead of food, integer because changing
+	 * using food as the key would require changing the hashcode, and equals method
+	 * in FoodData so contains is able to work. Possible to change if we absolutely need to though.
```
This comment needs to be updated to reflect your latest changes

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,92 @@
+	}
+	
+	
+	/* This method adds all the food in bill to the hashmap by checking the food name
```
```suggestion
	/* This method adds all the food in bill to the hashmap by checking the FoodData
```

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,92 @@
+	
+	// getter method for sales hashmap
+	public HashMap<FoodData, Integer> getSales(){
+		HashMap<FoodData, Integer> copySales = new HashMap<>(this.sales);
```
This one also just returns a shallow copy. To return deep copy, it will have to iterate through the map and put new pairs of FoodData and Integer

rdeleon-bran replied:
changed it

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,92 @@
+	}
+	
+	/* This method is sorted based off the most Sales */
+	public String  sortMostSales(){
```
I think this one should return an ArrayList<FoodData> so that higher-level class like Restaurant or View can do whatever they want with it.

In that case, make sure to return a deep copy of the array list

rdeleon-bran replied:
added

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,92 @@
+	
+	/* This method will sort the items based on how much money has been made off each
+	 * (most money to least) */
+	public String sortOffRevenue(){
```
I think this one should return an ArrayList<FoodData> so that higher-level class like Restaurant or View can do whatever they want with it.

In that case, make sure to return a deep copy of the array list

rdeleon-bran replied:
added

rdeleon-bran committed:
Update csc_335_restaurant/src/model/Bill.java

Co-authored-by: Bao Nguyen <88011047+BaoNguyen09@users.noreply.github.com>

rdeleon-bran committed:
Update csc_335_restaurant/src/model/Sales.java

Co-authored-by: Bao Nguyen <88011047+BaoNguyen09@users.noreply.github.com>

rdeleon-bran committed:
Made changes to Bill and Sales

rdeleon-bran committed:
changed the tests

rdeleon-bran committed:
Update FoodData.java

BaoNguyen09 commented on csc_335_restaurant/src/model/Bill.java:
```diff
@@ -33,17 +33,21 @@ public Bill(Bill other) {
 	
-	/* Get the tip */
-	public double getTip() {
-		return tip;
```
why do you delete this method? I think we would need to use it somewhere, currently i'm using it in Group class

rdeleon-bran replied:
mistake, just added it back.

BaoNguyen09 reviewed (changes requested):
Some final reviews

rdeleon-bran committed:
Update Bill.java

rdeleon-bran committed:
Update Bill.java

Tate-Smith commented on csc_335_restaurant/src/model/Bill.java:
```diff
@@ -33,25 +33,39 @@ public Bill(Bill other) {
 	
+	/* This method calculates the cost of only the food */
+	private void calculateFoodCost() {
+		double cost = 0;
```
make sure to use most up to date bill class

rdeleon-bran committed:
Update BillTests.java

rdeleon-bran committed:
Update FoodData.java

rdeleon-bran committed:
Update FoodDataTests.java

rdeleon-bran committed:
Update Sales.java

problem in the addOrder()

BaoNguyen09 committed:
fix: getPrice originally used is from FoodData so it depends on qty

rdeleon-bran committed:
Update Sales.java

rdeleon-bran committed:
Update FoodData.java

rdeleon-bran committed:
Update FoodDataTests.java

rdeleon-bran committed:
Update SalesTests.java

rdeleon-bran committed:
Add copyConstructor to food class

rdeleon-bran committed:
Update for Copy constructor

BaoNguyen09 commented on csc_335_restaurant/src/model/FoodData.java:
```diff
@@ -21,50 +22,48 @@ public FoodData(String name, FoodType type, double price, int quantity, String m
 	@Override
 	public double getPrice() {
-		return quantity * super.getPrice();
+		return super.getPrice();
```
If you do it like this, then it's equivalent to not having this override method

rdeleon-bran replied:
yes the Overrided method isn't necessary since it's not used at all in code

BaoNguyen09 replied:
then it's cleaner to just delete this method since it will inherit from Food anyway

BaoNguyen09 replied:
but i think this method is best overriden and reflect the price of FoodData (price * qty)

rdeleon-bran replied:
Overriding it doesn't make sense, just made a new method getTotalPrice() that does what the ovveride would do.

BaoNguyen09 commented on csc_335_restaurant/src/model/FoodData.java:
```diff
@@ -21,50 +22,48 @@ public FoodData(String name, FoodType type, double price, int quantity, String m
-		final int prime = 31;
-		int result = prime + Objects.hash(modifications, quantity);
-		return result;
+		return Objects.hash(super.hashCode());
```
i think the hashcode should be reverted cuz this makes FoodData no difference than Food

rdeleon-bran replied:
changed it back

BaoNguyen09 commented on csc_335_restaurant/src/model/Sales.java:
```diff
@@ -1,5 +1,99 @@
+	 *  */
+	public void addCompletedOrder(Bill bill) {
+		for (FoodData food: bill.getOrder()) {
+			Food copyFood = new Food(food.getName(), food.getType(), food.getPrice());
```
This is a much simpler method that don't require to change FoodData
```suggestion
			Food copyFood = (Food) food;
```

rdeleon-bran committed:
Update csc_335_restaurant/src/model/Sales.java

Co-authored-by: Bao Nguyen <88011047+BaoNguyen09@users.noreply.github.com>

rdeleon-bran committed:
Update FoodData.java

reverted HashCode and equals, also removed @Override from getPrice()

BaoNguyen09 commented on csc_335_restaurant/src/test/SalesTests.java:
I think there's one wrong test case

rdeleon-bran replied:
which one?

rdeleon-bran committed:
Update FoodData.java

Changed the Override of getPrice() to its own seperate method, getTotalPrice().

rdeleon-bran committed:
Update FoodDataTests.java

BaoNguyen09 committed:
fix: debug tests and code for Sales class

BaoNguyen09 committed:
feat: add more methods for Bill and Customer to handle payment

BaoNguyen09 reviewed (approved):
LGTM!
